### ISSUE TITLE
openssl rand accepts provider options but never uses them

### ISSUE THREAD

**Issue by simo5** [openssl/openssl#20850]
rand_main parses provider options via opt_provider(), however it never fetches a random provider, it just unconditionally uses the default.

**Comment by t8m** [comment]
I assume this is because the EVP_RAND is initialized before the potential rand provider is loaded.

**Comment by t8m** [comment]
@simo5 What does the random provider actually provide? Which algorithm? How do you call the rand command?

**Comment by simo5** [comment]
I called the algorithm PKCS11-RAND because for pkcs11 you have no idea what algorithm there is, it's like the rdrand engine.

But that is not the point, the openssl command simply ignore all provider related options.

I think we should look closely at the whole abstraction built around the rand provider in any case. It works if you are 

Other things I found: the provider loader API considers locking functions optional, yet if you do not provide them you do not go far, as early on the code requires them for seeding the "primary" drbg, which is not a thing in a hardware RNG (which, btw does not allow (re)seeding with the token I have).

I think you should try to convert the rdrand engine in a provider and then you'll figure out all the things that are not really right ...

OTOH, you may consider that I should only use HW sources for seeding and not for final numbers, but that doesn't work if your provider is a FIPS provider and you want random numbers to come out only from a certified source and not from some additional PRNG that is not certified.

Anyway I tried to call openssl rand -provider pkcs11 -propquery provider=pkcs11 1
And saw that the provider's rng code was untouched.

To force the use of the provider code for testing I now change the openssl.conf file to add a random section wth random = PKCS11-RAND and that does request bytes from the provider.

**Comment by paulidale** [comment]
The code is still going to _fetch_ CTR-DRBG regardless of the provider command line options (which are working BTW).  Using config and setting random = PKCS11-RAND changes what is being _fetched_ and thus works.

To address the real problem, what the rand command needs is another option to specify the RNG to be used along with any parameters it might need (a `-randopt` like `-pkeyopt` for genpkey).  Given these it should then call `RAND_set_seed_source_type()` and/or `RAND_set_DRBG_type()` to specify the RNG to use.  Alternatively, it could call `RAND_set0_public()` to override the public DRBG with the specified source.

**Comment by paulidale** [comment]
I edited the title to be more accurate.

**Comment by t8m** [comment]
Yeah, I assume the propquery is not applied to the CTR-DRBG fetch.

**Comment by paulidale** [comment]
Thanks for the extra investigation.

**Comment by paulidale** [comment]
The `-file` command line option isn't ideal either.

**Comment by minfrin** [comment]
Was looking for a canonical example of how to use a provider to generate random numbers (and be sure the provider generated those random numbers), and found myself here.

**Comment by bukka** [comment]
I guess the easiest way to verify it is using provider logs. You can check out https://github.com/bukka/osrand-provider and the tests in it.

**Comment by kgraefe** [comment]
You have to create an OpenSSL config file and set  the environment variable `OPENSSL_CONF` to its path:

```ini
openssl_conf = openssl_init

[openssl_init]
providers = providers
random = random

[providers]
default = default_provider
pkcs11 = pkcs11_provider

[default_provider]
activate = 1

[pkcs11_provider]
module = /path/to/pkcs11-provider.so
# any parameters your provider might have
activate = 1

[random]
random = PKCS11-RAND
```

Running `openssl rand -hex 10` with that draws 10 random bytes from your provider.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 2 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity for more than 2 months.
